[Rishit02] iP

Rishit02 opened the pull request:
### Riri

> **Not** from the creators of Siri 🙃 

Riri frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use
- All you need to do is,

1. download it from [here](https://github.com/Rishit02/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the chatbot constructor:

```
    public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage();
        try {
            tasks = new TaskList(storage.load());
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Rishit02 committed:
Rename, Greet and Exit

Rishit02 committed:
Echo

Rishit02 committed:
Revert "Echo"

This reverts commit d4c31097ddd7dfab9af0b562797919edcba9d27b.

Rishit02 committed:
Echo

Rishit02 committed:
Add, list

Rishit02 committed:
Mark as done

Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

Rishit02 committed:
Level 4. ToDos, Events, Deadlines

Add support for tracking three types of tasks:
ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

Rishit02 committed:
A-TextUiTesting

Use the input/output redirection technique to semi-automate the testing of Duke.

Rishit02 committed:
Added exceptions

Rishit02 committed:
Save data to hard disk

Write data to file.

The data needs to be saved to a file at the end of the operation so that it can be accessed later.

As a step towards using this data later, data has been saved after end of operation of chatbot.

Rishit02 committed:
Save function

The save function is completed.

The save function allows for users to load up their previously saved tasks.

The save function saves to a file called riri.txt in a directory called data. It then loads up the data if the file exists, else it creates a new list.

Rishit02 committed:
Merge pull request #1 from Rishit02/branch-Level-7

Branch level 7

Rishit02 committed:
Dates and Times

Read string and parse it as Java LocalDate objects.

Rishit02 committed:
Added formatters

Rishit02 committed:
Merge pull request #2 from Rishit02/branch-Level-8

Branch level 8

Rishit02 committed:
More OOP

Abstracted out the functionalities of the chatbot, the task list and the storage function.

Rishit02 committed:
Merge pull request #3 from Rishit02/AddIncrements

More OOP

Rishit02 committed:
Update delete function in tasklist

Rishit02 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Rishit02 committed:
Update TaskList

Rishit02 committed:
Use Gradle

Gradle is used. Checkstyle capabilites added.

Rishit02 committed:
Add JUnit tests

Add JUnit tests.

Test the construction of todo class and deadline class to ensure that tasks created are parsing strings properly.

Rishit02 committed:
Package the App as a JAR file

JukeBoxK commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,41 @@
+     * Function to help check if a task is done
+     * @return true if task is done, false otherwise
+     */
+    public boolean done() {
```
Perhaps changing to the name to something like `getIsDone` might be better, since this is a method, which typically have a verb for the name.

JukeBoxK commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,100 @@
+     * Introduces our chatbot and prompts the user to start asking questions
+     * @param name is the name given to our chatbot
+     * */
+    public void introduction(String name) {
```
Don't you think it might be better if the method is `introduce` or `start`?

JukeBoxK commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,100 @@
+    public void introduction(String name) {
+        System.out.println("--------------------------------------------------");
+        System.out.println("What's up. I'm " + name + ".");
+        System.out.println("I'm about to blow your world.");
```
What does it mean by "blow your world" ? :monocle_face:

JukeBoxK commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,63 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Do you think it might make the code clearer if you put some line break to separate out the different type of import statements?

JukeBoxK reviewed (commented):
Overall, except for a few method naming problem, the code looks great to me. It is simple but gets the job done perfectly.

doyelee0313 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,100 @@
+            }
+        }
+
+        throw new IllegalArgumentException("Unable to parse the date string using any of the specified patterns.");
```
Should it be better to inform the user about the date format?

doyelee0313 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,63 @@
+        this.by = byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+    }
+    private LocalDate parseDate(String date) {
+        DateTimeFormatter formatter1 = DateTimeFormatter.ofPattern("M/d/yyyy HHmm");
```
I like how you try to accept diverse styles of data format!

doyelee0313 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,56 @@
+        super(task);
+        this.fromDate = from;
+        this.toDate = to;
+        this.from = fromDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```
Would it be clearer to change this.from to fromDateFormatted?

doyelee0313 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,56 @@
+        this.fromDate = from;
+        this.toDate = to;
+        this.from = fromDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+        this.to = toDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```
Would it be clearer to change this.to to toDateFomatted?

doyelee0313 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,106 @@
+     * @param contents tasks to be added to task list
+     */
+    public TaskList(ArrayList<String> contents) {
+        for (String c : contents) {
```
Variable c can be written more clearer, but I think since it is only used here so should be fine.

Rishit02 committed:
Improve coding standard

Naming conventions are now followed. Changed function names as well as variable names to be more intuitively understood.

Rishit02 committed:
Create javadocs.

Add header comments to at least half of the non-private classes/methods.

Rishit02 committed:
Implement searchTasks method

The searchTasks method has been added to the TaskManager class to facilitate
efficient searching for tasks based on a specified keyword within their
descriptions. This feature addresses the need for users to quickly locate
relevant tasks, enhancing the overall usability of the application.

Rishit02 committed:
Merge branch 'branch-A-CodingStandard'

Rishit02 committed:
Merge branch 'branch-Level-9'

Rishit02 committed:
Add Checkstyle Plugin for Code Quality Checks

Introduce the Checkstyle plugin to the project to enforce coding standards
and improve code quality. This commit includes the necessary configuration
files and integrates Checkstyle into the build process.

Rishit02 committed:
Add Checkstyle Plugin for Code Quality Checks

Introduce the Checkstyle plugin to the project to enforce coding standards
and improve code quality. This commit includes the necessary configuration
files and integrates Checkstyle into the build process.

### Why It's Added

- Enhances code consistency and readability by enforcing a set of coding
  standards.
- Identifies potential issues early in the development process, reducing
  the likelihood of bugs and improving maintainability.
- Contributes to a more robust and professional codebase.

### What Is Included

- Integration of the Checkstyle plugin into the build process.
- Addition of Checkstyle configuration files tailored to the project's
  coding standards.

### How to Use

- Developers can now run Checkstyle checks using the appropriate build
  command or IDE integration.
- Violations of coding standards will be reported, and developers can
  address them to maintain a high code quality standard.

### Any Other Relevant Info

- Ensure that Checkstyle is configured according to the project's
  conventions and requirements.
- Refer to the project's documentation for guidelines on resolving
  Checkstyle violations.

Rishit02 committed:
Merge branch 'master' of https://github.com/Rishit02/ip

Rishit02 committed:
Enhance Date Parsing and User Communication

Improve the date parsing mechanism to handle various date formats more
robustly and enhance the communication with users regarding date-related
inputs. This commit aims to make date-related functionalities more
user-friendly and less error-prone.
What's Improved

1. **Date Parsing Mechanism:**
   - Refactor the date parsing logic to handle a wider range of date
     formats.
   - Enhance error handling for invalid date inputs, providing more
     informative error messages.

2. **User Communication:**
   - Improve user prompts and error messages related to date inputs.
   - Clearly communicate expected date formats to users for better
     understanding.

Why It's Improved

- Users may input dates in various formats, and the previous parsing
  mechanism had limitations.
- Enhancing user communication helps prevent date-related input errors,
  providing a more intuitive experience.

Rishit02 committed:
Improve User Input Handling Logic  Enhance the chatbot's user input processing logic to provide a more responsive and accurate interaction with users. This commit focuses on improving the understanding and handling of various user inputs for better overall chatbot performance.

### What's Improved

1. **Enhanced User Input Parsing:**
   - Refactor the user input parsing mechanism for improved accuracy.
   - Handle a broader range of user inputs with increased robustness.

2. **Contextual Understanding:**
   - Improve the chatbot's ability to understand user inputs within the
     context of the conversation.
   - Enhance contextual relevance for more natural and coherent responses.

3. **Error Handling:**
   - Implement better error handling for unexpected or ambiguous user
     inputs.
   - Provide clearer error messages to guide users in refining their
     queries.
### Why It's Improved

- Users may provide diverse inputs, and refining the input processing
  logic ensures a more seamless and user-friendly chatbot experience.
- Contextual understanding enhances the chatbot's ability to maintain
  coherent conversations with users.

### How It Benefits Users

- Users will experience a more intuitive interaction with the chatbot,
  as it can now better interpret a wider variety of inputs.
- Improved error handling contributes to a more user-friendly experience
  by guiding users in case of input issues.

### Any Other Relevant Info

- Ensure that the updated logic aligns with the chatbot's overall
  objectives and conversation flow.
- Thoroughly test the chatbot's responses with a variety of user inputs
  to verify the improvements.

Rishit02 committed:
Enhance Code Clarity with Additional Comments  Introduce comprehensive comments throughout the codebase to improve readability and provide valuable insights for fellow developers. This commit focuses on enhancing code documentation to facilitate better understanding and maintainability.

### What's Improved

1. **Code Comments:**
   - Add inline comments to explain complex or non-trivial sections of
     the code.
   - Provide clarifying comments for variable declarations, method
     invocations, and decision points.

2. **Documentation Updates:**
   - Enhance Javadoc comments for public methods, specifying their
     functionality and expected inputs/outputs.
   - Review and update existing comments to ensure accuracy and
     relevance.

### Why It's Improved

- Improved code documentation contributes to a more collaborative
  development environment, where team members can quickly grasp the
  purpose and functionality of various code components.
- Increased clarity in comments reduces the learning curve for new
  developers joining the project.

### How It Benefits Developers

- Developers can now navigate and understand the codebase more
  efficiently.
- Comprehensive comments serve as valuable documentation, aiding in
  troubleshooting, debugging, and future enhancements.

### Any Other Relevant Info

- Maintain consistency in commenting style and conventions throughout
  the codebase.
- Encourage ongoing documentation practices to keep the codebase
  well-documented and accessible.

Rishit02 committed:
Add JavaFX Dependencies to Build Configuration

Integrate JavaFX into the project by adding the necessary dependencies to
the Gradle build file. This commit sets up the build configuration to
include JavaFX libraries, enabling the development of JavaFX-based
applications.

The current situation:
- The project lacks JavaFX dependencies in the build configuration.

Why it needs to change:
- JavaFX is essential for developing graphical user interfaces in Java
  applications.

What is being done about it:
- Configure dependencies through Gradle for 'javafx-base,' 'javafx-controls,'
  'javafx-fxml,' and 'javafx-graphics.'
- Specify the JavaFX version (17.0.7) and classifier for different
  operating systems (win, mac, linux).

Why it is done that way:
- Configuring dependencies through Gradle simplifies the management of
  external libraries.

Any other relevant info:
- Verify that the specified JavaFX version and classifiers align with
  the project requirements.
- Additional JavaFX modules can be added as needed for specific project
  functionalities.

Rishit02 committed:
Refactor: Adding Custom Behavior to DialogBox in JavaFX  In this commit, I've refactored the JavaFX application to enhance the behavior of the DialogBox component. By adding custom functionality, we've improved the user interaction within the application.  The changes include incorporating specific actions and features into the DialogBox to better align with the application's requirements. These modifications aim to provide a more intuitive and responsive user experience.  For a detailed overview of the custom behavior implemented, please review the corresponding section in the codebase. Feel free to reach out if you have any questions or need further clarification on these changes.

Rishit02 committed:
Refactor GUI for Improved Maintainability and Readability

Refactor GUI for Improved Maintainability and Readability

The application's prototype is fully functional, but we've identified
several key issues that hinder the development process and code quality.
This commit addresses the following concerns:

1. **GUI Visual Enhancements:**
   - The GUI design process is streamlined by transitioning to FXML, an
     XML-based language for defining user interfaces.
   - Eliminate the need for constant rebuilds by leveraging FXML's
     concise syntax for UI definitions.

2. **Reduced Component Dependency:**
   - Components are decoupled to minimize interdependence. For instance,
     the Main class no longer needs to know specific details about the
     DialogBox or its internal components.
   - Enhance flexibility for future changes without risking the
     breakage of existing features.

3. **Tidier and Modular Code:**
   - Break down the monolithic Main class into modular components,
     separating visual tweaks, listeners, and utility methods.
   - Organize code logically to improve readability and simplify future
     maintenance tasks.

4. **Transition to FXML:**
   - Introduce FXML for UI definition to achieve a more concise and
     readable representation of the interface.
   - Example FXML snippet provided, showcasing the simplicity compared
     to the Java version.

- Improved GUI design process with FXML reduces the time and effort
  required for visual enhancements.
- Reduced component dependency enhances code modularity and flexibility.
- Tidier, modular code enhances readability, making it easier to locate
  and modify existing code.

- Ensure compatibility with the existing functionality while making
  these improvements.
- Encourage the use of FXML for future UI modifications and additions.
- Regularly review and refine the code organization to maintain
  cleanliness and modularity.

Rishit02 committed:
UI responds to user commands. Chat function allows for Duke to create tasks and respond to user appropriately.

Rishit02 committed:
Use assert feature

Use Assert Feature for Documenting Assumptions

This commit introduces the assert feature to document and validate
important assumptions within the codebase. Unlike JUnit assertions,
these assert statements serve as documentation for critical conditions
that should hold true at various points in the code.

Key Changes:
1. **Integration of Assert Statements:**
   - Assert statements have been strategically added to emphasize
     critical assumptions and conditions within the code.

2. **Documentation Through Assertions:**
   - Assertions are utilized as a form of inline documentation,
     making explicit the assumptions that are expected to be true at
     specific code locations.

3. **Benefits of Using Asserts:**
   - Enhances code readability by explicitly stating assumptions.
   - Serves as a tool for future developers to understand critical
     conditions that need to be maintained.

Rishit02 committed:
Improve code quality

This commit focuses on a critical examination of the existing codebase to enhance overall code quality. Several refactorings and improvements have been made to address issues related to readability, maintainability, and overall software design. The changes aim to create a cleaner, more organized, and efficient codebase.

Rishit02 committed:
Merge pull request #5 from Rishit02/branch-A-Assertions

Use assert feature

Rishit02 committed:
Merge branch 'master' into branch-A-CodeQuality

Rishit02 committed:
Add C-BetterSearch feature

Introduce the C-BetterSearch feature, enhancing the search functionality for items in the application. This improvement allows for more flexibility in searching, enabling the discovery of items even when the keyword matches only partially. The implementation provides a versatile search mechanism, contributing to a more user-friendly and accommodating user experience.

Rishit02 committed:
Refine Profile Picture Display

Rishit02 committed:
Improve GUI Aesthetics

Fine-tune padding, fonts, colors, and alignments to enhance the visual appeal of the user interface, providing a more pleasing and polished look.

Rishit02 committed:
Add color

Rishit02 committed:
Add Welcome Message and Aesthetic Improvements

Introduce a welcoming message in the GUI to enhance the user experience.
Apply color enhancements to improve the aesthetic appeal of the application.

Rishit02 committed:
Add user guide for task management

Introduce a comprehensive user guide providing instructions on how to perform various tasks in Riri, such as adding deadlines, events, deleting tasks, marking tasks as done or not done, and exiting the program. The user guide enhances user understanding and engagement with Riri's features.

Rishit02 committed:
Add save task feature

Introduce the save task feature to the user through the user guide.

Rishit02 committed:
Add Ui.png

Include a screenshot of the Riri GUI as Ui.png, providing users with a visual reference to the chatbot interface. Enhances documentation and helps users familiarise themselves with the application's appearance.

Rishit02 committed:
Refactor: Enhance code quality and readability

Performed code quality improvements, including refactoring methods, adding comprehensive JavaDoc comments, and ensuring consistent coding standards. The changes aim to enhance readability, maintainability, and overall code quality.

Rishit02 committed:
Update build.gradle

Update build.gradle to create workable *.jar files.

Rishit02 committed:
Smoke test

Rishit02 committed:
Update README.md
